File: aula04 23_09/Java/Padrao_Singleton.java
class Singleton {
    private static Singleton instance;

    // Construtor privado para impedir múltiplas instâncias
    private Singleton() {
        System.out.println("Criando a nova instância Singleton.");
    }

    // Método para obter a única instância
    public static Singleton getInstance() {
        if (instance == null) {
            instance = new Singleton();
        }
        return instance;
    }
}

// Exemplo de uso
public class Padrao_Singleton {
    public static void main(String[] args) {
        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();

        System.out.println(s1 == s2);  // True, pois ambas as instâncias são iguais
    }
}

File: aula04 23_09/Python/Padrao_Singleton.py
class Singleton:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Singleton, cls).__new__(cls)
            print("Nova instância criada.")
        return cls._instance

# Exemplo de uso
if __name__ == "__main__":
    s1 = Singleton()
    s2 = Singleton()

    print(s1 is s2)  # True, ambas as instâncias são a mesma
